mLRS/CommonTx/esp-jr_pin5_interface.h

```diff
@@ -0,0 +1,147 @@
+    void uart_rx_callback(uint8_t c);
+    void uart_tc_callback(void);
+
+    // parser
+    typedef enum {
+        STATE_IDLE = 0,
+
+        // mBridge receive states
+        STATE_RECEIVE_MBRIDGE_STX2,
+        STATE_RECEIVE_MBRIDGE_LEN,
+        STATE_RECEIVE_MBRIDGE_SERIALPACKET,
+        STATE_RECEIVE_MBRIDGE_CHANNELPACKET,
+        STATE_RECEIVE_MBRIDGE_COMMANDPACKET,
+
+        // CRSF receive states
+        STATE_RECEIVE_CRSF_LEN,
+        STATE_RECEIVE_CRSF_PAYLOAD,
+        STATE_RECEIVE_CRSF_CRC,
+    } STATE_ENUM;
+#define STATE_TRANSMIT_START STATE_IDLE // Don't need this state for full duplex
+
+    // not used in this class, but required by the children, so just add them here
+    // no need for volatile since used only in isr context
+    uint8_t state;
+    uint8_t len;
+    uint8_t cnt;
+    uint16_t tlast_us;
+
+    // check and rescue
+    // the FRM303 can get stuck, whatever we tried, so brutal rescue
+    // can't hurt generally as safety net
+    uint32_t nottransmiting_tlast_ms;
+    void CheckAndRescue(void);
+};
+
+
+void tPin5BridgeBase::Init(void)
+{
+    state = STATE_IDLE;
+    len = 0;
+    cnt = 0;
+    tlast_us = 0;
+
+    telemetry_start_next_tick = false;
+    telemetry_state = 0;
+
+    nottransmiting_tlast_ms = 0;
+    uart_init();
+}
+
+
+void tPin5BridgeBase::TelemetryStart(void)
+{
+    telemetry_start_next_tick = true;
+}
+
+
+//-------------------------------------------------------
+// Interface to the uart hardware peripheral used for the bridge
+// called in isr context
+
+void tPin5BridgeBase::pin5_tx_enable(bool enable_flag)
+{
+    // nothing to do for full duplex
+}
+
+
+// we do not add a delay here before we transmit
+// the logic analyzer shows this gives a 30-35 us gap nevertheless, which is perfect
+
+void tPin5BridgeBase::uart_rx_callback(uint8_t c)
+{
+    // Not yet used
+    // parse_nextchar(c);
+}
+
+
+void tPin5BridgeBase::uart_tc_callback(void)
+{
+    // Not yet used
+    // pin5_tx_enable(false); // switches on rx
+    //state = STATE_IDLE;
+}
+
+
+//-------------------------------------------------------
+// Check and rescue
+// a good place to call it could be ChannelsUpdated()
+// Note: For the FRM303 it was observed that the TC callback may be missed in the uart isr, basically when
+// the jrpin5's uart isr priority is too low. This caused the jrpin5 loop to get stuck in STATE_TRANSMITING,
+// and not even channel data would be received anymore (= very catastrophic). This code avoids this.
+// With proper isr priorities, the issue is mainly gone, but the code remains, as safety net.
+//
+// For full duplex CRSF (internal module), this is used to poll for received bytes and pace telemetry
+
+void tPin5BridgeBase::CheckAndRescue(void)
+{
+    while (uart_rx_available()) { // polling for now
+        parse_nextchar(uart_getc());
+    }
+
+    uint32_t tnow_ms = millis32();
+
+    // Send telemetry and mbridge at the same rate as external module to be sure lua can keep up
```

Comment:
from scratch this owuld worry me
doesn't it assume that the while loop before really stops at the end of a frame, so this transmit starts comes in sync?

generally I also try to not to do too much in a sub loop, to give the higher loop a chance to run. Am wondering also from this point if the while loop before shouldn't have a test for an end of a framre

Reply:
I believe there is no need to be in sync as this is a full duplex serial connection.  We are doing two independent things here:  Gathering up all the received data and sending if it's time.

Reply:
would this be even true if in the parse loop two frames would be parsed?

Reply:
Why wouldn't it be true? I admit I haven't looked at the EdgeTx side to see if they made any stupid assumptions, but why would it matter?  This is full duplex.  We can send our telemetry to the radio whenever we like.  I choose to send it with about 4ms between each message to give the Lua script the same amount of time to process a message as it had in the half duplex case.

Reply:
for a god number of frames there will be a response, but if it parses two messages and call transmit only once a response can be missed (if I recall the code doesn't buffer responses)

Reply:
OK.  I'm not sure if that actually ever happens, but it's a good point and another good reason to stick with 400K baud.
I'll rework this.